### PR TITLE
Avoid invalid cast exception in BindImplicitConstructorInitializer for top level statements with non-object base type.

### LINKED ISSUE #70909: "csc.exe" exited with code -2146232797, crash in BindImplicitConstructorInitializer
**Version Used**: .NET 8.0.100 + MSBuild version 17.8.3 (also tried lower version, same story)

**Steps to Reproduce**:

**1)** Clone https://github.com/iXab3r/DotnetCSCTopLevelStatementsError
**2a)** ` dotnet build` Try to build (CSC will crash)
**OR** 
**2b)** `dotnet test` Run tests (**[ShouldNotThrowOnCompilation](https://github.com/iXab3r/DotnetCSCTopLevelStatementsError/blob/1ed0fc8ec38d1f588b7d9648a33144cf61cd3ef3/CSCCompileTimeError.Tests/Tests.cs#L12C26-L12C26)** throws)

When compiling a program with top-level statements and there is a code that extends Program, compiler throws (both CSC and MethodCompiler). 
**Note:** problem reproduces only when class that Program is extending is in a separate file, at least that is what tests are showing

**Expected Behavior**:
Code compiles

**Actual Behavior**:
2a - **CSC** crashes with error -2146232797
2b - **compilation.GetDiagnostics** throws an exception 
```
System.InvalidCastException: Unable to cast object of type 'Microsoft.CodeAnalysis.CSharp.Syntax.CompilationUnitSyntax' to type 'Microsoft.CodeAnalysis.CSharp.Syntax.BaseTypeDeclarationSyntax'.
```

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #70909.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 1]
```diff
@@ -3781,17 +3781,16 @@ internal virtual BoundExpressionStatement BindConstructorInitializer(Constructor
...
                 {
-                    outerBinder = binderFactory.GetInTypeBodyBinder(typeDecl);
-                }
-                else
-                {
-                    SyntaxToken bodyToken = GetImplicitConstructorBodyToken(containerNode);
-                    outerBinder = binderFactory.GetBinder(containerNode, bodyToken.Position);
+                    // Must be a source of top level statements with a partial type declaration
+                    // that specifies non-object base. The object base is handled above.
+                    containerNode = containingType.DeclaringSyntaxReferences.Select(r => r.GetSyntax()).OfType<TypeDeclarationSyntax>().First();
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:3789]: This comment implies that we need to be filtering for the type declaration that actually has the base type declaration, which is not what this code actually does. Is it worth adding another test for a `Program` with 3 parts: the top-level statement, an empty class body, and a body that declares the base type?

Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:3789]: > The comment explains under what conditions we can get here. It doesn't say what to do, or what filtering to apply.

That isn't how I'm reading it though. I find it confusing and I think we need to change the wording to be clearer. My suggestion would be to add a sentence that states something like `We need an actual TypeDeclarationSyntax in order to locate the correct binder for this case.`

**Review by 333fred — APPROVED** [review]
LGTM with a comment reword

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review

**Pull request merged**